File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.js
function solution(triangle) {
    const dp = Array.from({ length: triangle.length }, (_, i) => Array(i + 1).fill(0));
    dp[0][0] = triangle[0][0];
    
    for (let i = 1; i < triangle.length; i++) {
        for (let j = 0; j < triangle[i].length; j++) {
            if (j - 1 >= 0)
                dp[i][j] = Math.max(dp[i][j], triangle[i][j] + dp[i - 1][j - 1]);
            if (j < i)
                dp[i][j] = Math.max(dp[i][j], triangle[i][j] + dp[i - 1][j]);
        }
    }
    return Math.max(...dp[dp.length - 1]);
}

File: 백준/Silver/1932. 정수 삼각형/정수 삼각형.py
import sys
input = sys.stdin.readline

N = int(input())
triangle = []

for _ in range(N):
  triangle.append(list(map(int, input().split())))

dp = [[0] * (i + 1) for i in range(N)]
dp[0][0] = triangle[0][0]

for i in range(1, N):
  for j in range(i + 1):
    if j - 1 >= 0:
      dp[i][j] = max(dp[i][j], triangle[i][j] + dp[i - 1][j - 1])
    if j < i:
      dp[i][j] = max(dp[i][j], triangle[i][j] + dp[i - 1][j])

print(max(dp[-1]))